File: src/solutions/2025/03/zig/main.zig
const std = @import("std");
const aoc = @import("aoc");

const options = aoc.SolutionOptions.init();

fn solve(input: []const u8, is_test: bool, p2: bool) []const u8 {
    _ = is_test;
    var res: i64 = 0;

    var lines = std.mem.splitSequence(u8, input, "\n");
    while (lines.next()) |line| {
        var start: usize = 0;
        const max: usize = if (p2) 12 else 2;

        for (0..max) |digit| {
            var marked: usize = 0;
            var n: i64 = 0;
            const t = max - 1 - digit;
            const l = line.len - t;
            for (start..l) |it| {
                const parsed = line[it] - '0';
                if (n < parsed) {
                    marked = it;
                    n = parsed;
                }
            }

            start = marked + 1;
            res += n * std.math.pow(i64, 10, @intCast(t));
        }
    }

    const buf = std.heap.c_allocator.alloc(u8, 42) catch unreachable;
    const str = std.fmt.bufPrint(buf, "{}", .{res}) catch "";
    return str;
}

fn part1(input: []const u8, is_test: bool) []const u8 {
    return solve(input, is_test, false);
}

fn part2(input: []const u8, is_test: bool) []const u8 {
    return solve(input, is_test, true);
}

pub fn main() !void {
    try aoc.run(part1, part2, options);
}

File: src/solutions/2025/03/rust/main.rs
extern crate aoc_core;
use std::{env, iter::FromIterator};

fn get_options() -> aoc_core::options::SolutionOptions {
   aoc_core::options::SolutionOptions {
      ..Default::default()
   }
}

fn solve(input: &str, _is_test: bool, p2: bool) -> String {
   input
      .lines()
      .fold(0, |acc, line| {
         let mut res: u64 = 0;
         let mut start: usize = 0;
         let max: usize = if p2 { 12 } else { 2 };
         let bytes = line.as_bytes();
         for digit in 0..max {
            let mut marked = 0;
            let mut n: u64 = 0;
            let t = max - 1 - digit;
            let l = bytes.len() - t;
            for it in start..l {
               let parsed = (bytes[it] - b'0') as u64;
               if n < parsed {
                  marked = it;
                  n = parsed;
               }
            }
            start = marked + 1;
            res += n * (10_u64.pow(t as u32));
         }
         res + acc
      })
      .to_string()
}

fn part_1(input: &str, is_test: bool) -> String {
   solve(input, is_test, false)
}

fn part_2(input: &str, is_test: bool) -> String {
   solve(input, is_test, true)
}

fn main() {
   let args = Vec::from_iter(env::args());
   aoc_core::run(
      args.iter().map(|x| x.as_ref()).collect::<Vec<_>>(),
      part_1,
      part_2,
      get_options(),
   );
}
